plasmapy/tests/helpers/exceptions.py

```diff
@@ -0,0 +1,109 @@
+"""Exceptions that describe different types of test failures."""
+
+__all__ = [
+    "Failed",
+    "UnexpectedResultError",
+    "InconsistentTypeError",
+    "MissingExceptionError",
+    "UnexpectedExceptionError",
+    "MissingExceptionError",
+    "UnexpectedWarningError",
+    "MissingWarningError",
+    "InvalidTestError",
+    "ExceptionMismatchError",
+    "WarningMismatchError",
+]
+
+import pytest
+
+# TODO: get the error message while raising this to be simply "Failed"
+#       rather than plasmapy.tests.helpers.exceptions.Failed
+
+
+class Failed(pytest.fail.Exception):
+    """
+    Base exception for test failures.
+
+    Notes
+    -----
+    This exception was derived from `~pytest.fail.Exception`, which in
+    turn was derived from `BaseException` (not `Exception`).
+    """
+
+    pass
+
+
+class InconsistentTypeError(Failed):
+    """
+    Exception for when the type of the actual result differs from the
+    type of the expected result.
+    """
+
+    pass
+
+
+class MissingExceptionError(Failed):
+    """
+    Exception for when the expected exception is not raised.
+    """
+
+    pass
+
+
+class UnexpectedExceptionError(Failed):
+    """
+    Exception for when an exception is raised unexpectedly.
+    """
+
+    pass
+
+
+class ExceptionMismatchError(UnexpectedExceptionError, MissingExceptionError):
+    """
+    Exception for when an exception is expected, but a different
+    exception is raised.
+    """
+
+    pass
+
+
+class UnexpectedWarningError(Failed):
+    """
+    Exception for when a warning is issued unexpectedly.
+    """
+
+    pass
+
+
+class MissingWarningError(Failed):
+    """
+    Exception for when an expected warning is not issued.
+    """
+
+    pass
+
+
+class WarningMismatchError(UnexpectedWarningError, MissingWarningError):
+    """
+    Exception for when a warning is expected, but one or more
+    different warnings were issued instead.
+    """
+
+    pass
+
+
+class UnexpectedResultError(Failed):
+    """
+    Exception for when the actual result differs from the expected
+    result by more than the allowed tolerance.
+    """
+
+    pass
+
+
+class InvalidTestError(Exception):
```

Comment:
I'm assuming you're inheriting form `Exception`, and not `Failed`, on purpose?  This is a weird one since the test fails, but not due to the tested code failing.  Not sure which inheritance is more appropriate.

Reply:
Yeah...this is different than a test failure, because the test was not actually able to be performed.  It's kind of like how pytest treats errors and failures differently.  I've been using `Failed` to represent failures where the problem is with the production code, and `InvalidTestError` to represent errors where the problem is the test code.

Reply:
Is there currently a use case for this?  Does `pytest` already have something like this in it?  Just curious if we're duplicating functionality that already exists.

Reply:
The closest thing is a `UsageError` in pytest...though the docstring specifically refers to pytest so I don't think it's quite apropos to inherit from it:  
```Python
@final
class UsageError(Exception):
    """Error in pytest usage or invocation"""
```

Reply:
The use case is for a developer who is writing tests, but makes an error in setting the test up.  That has happened to me...dare I say...more than once.